common/src/character_controller/mod.rs

```diff
@@ -53,68 +56,190 @@ fn run_standard_character_step(
     ctx: &CharacterControllerContext,
     position: &mut Position,
     velocity: &mut na::Vector3<f32>,
+    on_ground: &mut bool,
 ) {
+    let mut ground_normal = None;
+    if *on_ground {
+        ground_normal = get_ground_normal(ctx, position);
+    }
+
+    // Handle jumping
+    if ctx.jump_input && ground_normal.is_some() {
+        let horizontal_velocity = *velocity - *ctx.up * ctx.up.dot(velocity);
+        *velocity = horizontal_velocity + *ctx.up * ctx.cfg.jump_speed;
+        ground_normal = None;
+    }
+
     let old_velocity = *velocity;
 
     // Update velocity
-    let current_to_target_velocity = ctx.movement_input * ctx.cfg.max_ground_speed - *velocity;
-    let max_delta_velocity = ctx.cfg.ground_acceleration * ctx.dt_seconds;
-    if current_to_target_velocity.norm_squared() > math::sqr(max_delta_velocity) {
-        *velocity += current_to_target_velocity.normalize() * max_delta_velocity;
+    if let Some(ground_normal) = ground_normal {
+        apply_ground_controls(ctx, &ground_normal, velocity);
     } else {
-        *velocity += current_to_target_velocity;
+        apply_air_controls(ctx, velocity);
+
+        // Apply air resistance
+        *velocity *= (-ctx.cfg.air_resistance * ctx.dt_seconds).exp();
     }
 
+    // Apply gravity
+    *velocity -= *ctx.up * ctx.cfg.gravity_acceleration * ctx.dt_seconds;
+
+    // Apply speed cap
+    *velocity = velocity.cap_magnitude(ctx.cfg.speed_cap);
+
     // Estimate the average velocity by using the average of the old velocity and new velocity,
     // which has the effect of modeling a velocity that changes linearly over the timestep.
     // This is necessary to avoid the following two issues:
     // 1. Input lag, which would occur if only the old velocity was used
     // 2. Movement artifacts, which would occur if only the new velocity was used. One
     //    example of such an artifact is the character moving backwards slightly when they
     //    stop moving after releasing a direction key.
-    let estimated_average_velocity = (*velocity + old_velocity) * 0.5;
+    let average_velocity = (*velocity + old_velocity) * 0.5;
 
-    apply_velocity(ctx, estimated_average_velocity, position, velocity);
+    // Handle actual movement
+    apply_velocity(
+        ctx,
+        average_velocity,
+        position,
+        velocity,
+        &mut ground_normal,
+    );
+
+    *on_ground = ground_normal.is_some();
```

Comment:
It's a bit difficult to follow the evolution of `ground_normal`. I wonder if we could factor most of these `&mut` parameters out into fields on a common struct that these are all methods on?

Reply:
There do seem to be two competing programming principles that I haven't fully worked out:
- Bundling common things together into a struct to reduce parameter count and reduce the amount of code that needs to be read and understood
- Explicitly passing in parameters so that reading the code makes it more apparent exactly what each section's dependencies are.

For this PR, I ended up erring on the side of the second bullet point for all the mutable parameters, but I'm definitely not convinced that's the right move.

These parameters can definitely be factored out into a struct (player-state?). One potential pitfall is that `ground_normal` has to be an `Option`, so if there's a struct for ground-specific movement, it needs to either handle the `None` case (even though flow control would never reach there) or unwrap it. Honestly, I think that's fine, but we do lose some of the correctness benefits.

Reply:
Yeah, it can be awkward. We could have multiple structs (they're temporaries anyway) but it can get a bit boilerplatey. Not necessarily any action required here.

Reply:
Sounds good. In that case, I'll stick with the current implementation, but if you have any concrete suggestions here, I'm happy to try them out.